torchao/dtypes/uint2.py

```diff
@@ -0,0 +1,411 @@
+
+    #@torch.compile
+    def unpack_uint2(uint8_data: torch.Tensor) -> torch.Tensor:
+        # since we are using uint8 we will decode 4 entries per byte
+        shape = uint8_data.shape
+        uint8_data = uint8_data.to(torch.uint8)
+        first_elements = ((uint8_data >> 6) & 0b11)
+        second_elements = ((uint8_data >> 4) & 0b11)
+        third_elements = ((uint8_data >> 2) & 0b11)
+        fourth_elements = (uint8_data & 0b11)
+        return torch.stack((first_elements, second_elements, third_elements, fourth_elements), dim=-1).view(up_size(shape))
+
+    #packing uint8
+    #@torch.compile
+    def pack_uint2(uint8_data: torch.Tensor) -> torch.Tensor:
+        shape = uint8_data.shape
+        assert shape[-1] % 4 == 0
+        uint8_data = uint8_data.contiguous().view(-1)
+        packed_data = (uint8_data[::4] << 6 | uint8_data[1::4] << 4 | uint8_data[2::4] << 2 | uint8_data[3::4]).view(down_size(shape))
+        return packed_data
+
+
+def fill_defaults(args, n, defaults_tail):
+    """
+    __torch_dispatch__ doesn't guarantee the number of arguments you are
+    passed (e.g., defaulted arguments are not passed); but usually it is
+    convenient to pad out the arguments list with defaults.  This function
+    helps you do that.
+    Args:
+        args: the list of positional arguments passed to __torch_dispatch__
+        n: the number of arguments you are expecting to get
+        defaults_tail: default values for the arguments, starting from the
+            end of the list
+    Example:
+        >>> fill_defaults([1, 2, 3], 5, [3, 4, 5])
+        [1, 2, 3, 4, 5]
+        >>> fill_defaults([1, 2, 3], 5, [None, None, None])
+        [1, 2, 3, None, None]]
+    """
+    if n - len(defaults_tail) > len(args):
+        raise RuntimeError("not enough defaults to fill arguments")
+    r = list(args)
+    for i in range(len(args), n):
+        r.append(defaults_tail[i - n + len(defaults_tail)])
+    return r
+
+
+#qtensor_lib = Library("qtensors", "DEF")
+qtensor_lib.define(
+    "quantize_per_tensor_uint2(Tensor input, float scale, int zero_point) -> Tensor"
+)
+
+
+@impl(qtensor_lib, "quantize_per_tensor_uint2", "CompositeExplicitAutograd")
+def quantize_per_tensor_uint2(
+    input: torch.Tensor,
+    scale: float = 1.0,
+    zero_point: int = 1,
+) -> torch.Tensor:
+    inv_scale = 1.0 / scale
+    return pack_uint2(
+        torch.clamp(torch.round(input * inv_scale) + zero_point, 0, 2).to(torch.uint8)
+    )
+
+
+qtensor_lib.define(
+    "dequantize_per_tensor_uint2(Tensor input, float scale, int zero_point) -> Tensor"
+)
+
+
+@impl(qtensor_lib, "dequantize_per_tensor_uint2", "CompositeExplicitAutograd")
+def dequantize_per_tensor_uint2(
+    input: torch.Tensor,
+    scale: float = 1.0,
+    zero_point: int = 1,
+) -> torch.Tensor:
+    input = unpack_uint2(input)
+    return (input.view(torch.uint8).to(torch.float32) - zero_point) * scale
+
+
+class UInt2Tensor(torch.Tensor):
```

Comment:
maybe we want to generalize UInt4Tensor to work with uint1 to uint7 directly

Reply:
Good point, will generalize it.